### ISSUE TITLE
Consider expanding mopidy-local browse interface.

### ISSUE THREAD

**Issue by adamcik** [mopidy/mopidy#877]
https://github.com/tkem/mopidy-local-sqlite has been doing a great job of implementing a much fuller browse interface which covers ideas such as having a virtual file system based on different breakdowns.

Ideally we want this functionality for all local backend variants without each of them having to re-implement this fully. On way to achieve this would be to extend the current local browse interface with method(s) for querying the data that this "standard" browse implementation would need. This would be done by having the base library provide a fully fledged `browse` and adding the helpers as `return []` dummies. This would mean local library authors still have the choice of making their own browse if they want something else than the default, but for the common case they can just implement the helpers and easily get up and running.

@tkem any suggestions on how we should do this given your experience with mopidy-local-sqlite?

#562 also comes to mind as something which could be added to such a generic local browse.

**Comment by tkem** [comment]
Update: a couple of thoughts after supporting Mopidy-Local-SQLite for two more months:

- If Mopidy continues to support MPC's `list <type>`, there should be a corresponding `Library.list(type)` to support that efficiently. This is not necessarily limited to local libraries. Of course, backends like spotify or gmusic where listing all available artists/albums/whatever doesn't really make sense may choose to "opt out" of this, i.e. return `None`. Alternatively, support for MPC's `list` command should be dropped altogether. 

- Mopidy's `local.Library.add(track)` interface is essentially flawed, IMHO. It should either

  - provide unique album and artist URIs, resolving the issues brought up in https://discuss.mopidy.com/t/data-model-semantics-of-num-tracks-for-multi-disc-albums/526/2 (which involves deciding what *really* is a unique album and/or artist), or

  - just pass the scanner data as-is, leaving it all to the local library implementation

  The current specification and implementation are quite unsatisfactory, to say the least. On the one hand, the local lib doesn't have access to the original data (not just images, but also simple things such as the file's path, which has to be extracted from the track URI by reading the code -- the relationship between track URIs and file paths is not documented anywhere, AFAIK). On the other hand, it's passed a `Track` model where much of the relevant details have to be figured out/made up by the local lib, anyway (artist and album URIs, for example, not to mention the `num_disc/num_tracks` mess).

- If you really want to extend the current local browse interface so it becomes more like a "framework", it should definitely handle album/artist URI assignment. If something like a `Library.list()` method is implemented, this could serve to provide browse URIs, which can be handled by the base class. Also "file system browsing", as implemented in Mopidy-Local-SQLite v0.9.2, could be handled by Mopidy core, . However, I think all this would require passing *complete* `Track` models to the local library, i.e. handling album/artist URIs, `track_num/disc_num`, *and* images sensibly. If you don't see these things in Mopidy core, it's probably better to leave things as-is, and just give local libs better access to the scanner data...

**Comment by adamcik** [comment]
1. Agreed, we might be moving this into search by allowing you to limit by type. There is an open PR which is a step in this direction, but I haven't had time to review it properly yet

2. The last code I wrote before leaving on vacation was related to this. The new interface would be `library.add(track, tags=None, duration=None)`, but it might make sense just to reduce it to tags+duration and leave a util for the default conversion around.

3. I'm not sure what we want here yet, and as such the more faster paced experimentation mopidy-local-sqlite is doing is awesome for getting a better idea of what we might want. As for file system browsing I was wondering about adding something which just exposes directories without the need to prescan, but then with limited metadata until playback and only browse.

**Comment by tkem** [comment]
Ad 1: This should also take into account `mpc list genre`, `mpc list date` and whatever else Mopidy supports. So I don't know if integrating this with `search()` is such a good idea. Unless you want to add `SearchResult.genres`, etc. and/or make genres, dates, etc. proper Models😄

Maybe a new method for this that gets the type (as in query keys) and returns a list if unique values (as strings) would be more fitting? Don't know enough about the MPD protocol to say that serves the purpose, though.

**Comment by tkem** [comment]
Making this it's own method makes it also a little easier for back ends that don't support "list". I guess most back ends will support "search", but only few (mostly local ones) will or even can support "list".

**Comment by adamcik** [comment]
I was also considering if forcing it into browse could somehow make sense. But I'm not finding any nice way of doing it. Looking over the MPD protocol for `list` and `count` we need to support:

1. Listing all the unique values for given type
2. Possibly applying some filter to that listing
3. Possibly applying some grouping to that listing

To do this efficiently we likely need to pass both the filters and the groupings in to the API, and at this point having this in browse becomes messy to say the least. And forcing it into search also sounds questionable having looked the protocol and how for instance ncmpcpp uses it.

**Comment by tkem** [comment]
To get back to the original topic of this issue ("expanding mopidy-local browse interface"), after much pondering I'd like to answer with a counter question:

Does Mopidy *really* need a "local" backend built-in?

I know, this sounds scary ;-), but:

- Many (if not most) people still come to Mopidy because of its Spotify support. Some never, ever use the `local` backend, and some even disable it (tkem/mopidy-local-sqlite#30). Now, `mopidy-spotify` is an extension that has to be installed seperately, and `local` is built-in. Why?

- A "local" backend doesn't necessarily need to extend the `LocalLibrary` interface. Look at @rawdlite's mopidy-beets-local for example -- it's definitely "local", as it involves files stored on the local HDD/SD-card, but (AFAIK) no `LocalLibrary` interface, no `mopidy local scan`, etc. Sure, the mopidy-local framework gave me a head start when I implemented mopidy-local-sqlite, but the more time I spend on that, the more limiting this becomes (e.g. the local images/pass scanner data to `LocalLibrary.add()` issue). For future versions of mopidy-local-sqlite, I'm also thinking about bypassing `mopidy local scan`, for example periodically doing a scan in the background, or even something based on `inotify` (if available).

- Dropping the `local` backend and (more or less) leaving this to extension developers probably would help core developers to concentrate on functionality and bug fixes helping *all* extensions. I guess your (and @jodal's, and everybody else's who works on Mopidy core) precious time could thus be better spent. Let others deal with "my local files don't show up in the browse interface where I expect them to" issues ;-)

So, IMHO, instead of thinking about *expanding* the mopidy-local interface, my advice is to drop it altogether. Wrap up the `json` backend in an extension, similar to `mopidy-spotify`. Concentrate on core, the `stream` backend, and `mopidy.audio.scan` -- that's all it really needs to write a "local" extensions, AFAICS.

Well, maybe, keep (basic) local playlist support for now, but that remains to be discussed...

**Comment by adamcik** [comment]
I've actually been considering this myself as an alternative to trying to bring mopidy-local-sqlite into the main distribution. As for the other points:

- Mostly because in the beginning there was just spotify, then local or spotify and eventually multibackend support. Then spotify got kicked out for when making debian packages. In other words, there isn't any especially good reason any more. Other than the possible value of providing some things for free for local implementors, but that could easily be done externally like you've done for podcasts.

  For my own part this has also been somewhat shaped by previous experiences many years ago trying to get Amarok, MPD and various other players to work well with a 100k+ tracks in a central archive across 20+ machines. So the idea has always been to facilitate centralized scanning external from mopidy (say on the file server) and have the client machines connect to the central DB.

- `inotify` is a possibility, bet there are a lot of corner cases that can mess it up. So often you will end up having to reinitialize the entire watch setup, to the point that you might just be better off doing a `stat` on everything once in a while.

  As for the specific case of image data / raw access to tags, I stared a branch for that and have the code mostly done. I'll loop you in on the PR when I get it pushed out.

 For background scanning I've been trying to decide if we should provide this as a core service or not. Advantage is we wouldn't have random backends all start this in duplicate. The downside would be having to avoid resource starvation when say stream needs to do a lookup now to play something and a local scan is using all the "slots". So just having dedicated one in each backend might still be the same thing, and the cost of threads not doing anything should only be some wasted memory.

- As for moving the burden of making local well away from core this would be a huge win :-) Though part of be is tempted to supplement the stream backend with a filesystem backend. This backend would basically only implement browse for a configured directory and only fetch the metadata at lookup time like stream does - but could very well be that nobody really needs such a backend and that those that would are better served with some local variant.

So in summary I don't think this is unreasonable, even without having a filesystem backend to "replace" it.

**Comment by tkem** [comment]
Thanks for your elaborate response! Guess I'll have to digest that over the next couple of days, so for now just let me tell you that I had to use `inotify` for some work-related projects, so I'm pretty much aware of its pitfalls/shortcoming and I'm not taking it lightly ;-) It's just one possibility I'm considering for mopidy-local-sqlite.

**Comment by tkem** [comment]
As for your "filesystem" backend -- this could be a valuable extension, whether included with Mopidy "base" or not. I guess if (and only if) "local" is really dropped from the "base" package, the `json` backend might not get that much attention any longer. So, instead of providing the full-featured "json" backend, you might be better advised to provide something more basic as the *de-facto* standard local library provider.

And, yes, I feel flattered reading things like "bring mopidy-local-sqlite into the main distribution", but I think that would actually limit its possibilities/options for future development (or my desire to experiment).

**Comment by adamcik** [comment]
My biggest concern for splitting this out currently is actually that a lot of our "core" tests are in fact the integration tests that we have for local. So cleaning that up and replacing those tests with proper unittest for core would be required.

**Comment by adamcik** [comment]
Closing in favour of #1003 which opened for moving local "out" of mopidy.
